File: solutions/1442. Count Triplets That Can Form Two Arrays of Equal XOR/1442.cpp
// Time complexity: O(n)
// Space complexity: O(n)
class Solution {
 public:
  int countTriplets(const vector<int>& arr) {
    int ans = 0;
    int pref = 0;
    unordered_map<int, int> countMap = {{0, 1}};
    unordered_map<int, int> totalMap;

    for (int i = 0; i < arr.size(); ++i) {
      pref ^= arr[i];
      ans += countMap[pref]++ * i - totalMap[pref];
      totalMap[pref] += i + 1;
    }

    return ans;
  }
};

File: solutions/1442. Count Triplets That Can Form Two Arrays of Equal XOR/1442.py
# Time complexity: O(n)
# Space complexity: O(n)
class Solution:
  def countTriplets(self, arr: List[int]) -> int:
    ans = 0
    pref  = 0
    countMap = defaultdict(int, {0: 1})
    totalMap = defaultdict(int)

    for i, val in enumerate(arr):
      pref ^= val
      ans += countMap[pref] * i - totalMap[pref]
      countMap[pref] += 1
      totalMap[pref] += i + 1

    return ans